### PR TITLE
Segment Replication - Fixes for test SegmentReplicationIT.testReplicaHasDiffFilesThanPrimary

### LINKED ISSUE #6885: [BUG] Test org.opensearch.indices.replication.SegmentReplicationIT.testReplicaHasDiffFilesThanPrimary is flaky
**Describe the bug**
- `org.opensearch.indices.replication.SegmentReplicationIT.testReplicaHasDiffFilesThanPrimary` is flaky
- https://build.ci.opensearch.org/job/gradle-check/13054/testReport/org.opensearch.indices.replication/SegmentReplicationIT/testReplicaHasDiffFilesThanPrimary_2/

**To Reproduce**
```
./gradlew ':server:internalClusterTest' --tests "org.opensearch.indices.replication.SegmentReplicationIT.testReplicaHasDiffFilesThanPrimary" -Dtests.seed=40450506CE0A9D4A -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=es-PE -Dtests.timezone=Antarctica/Vostok -Druntime.java=19
```

**Screenshots**
- https://build.ci.opensearch.org/job/gradle-check/13054

**Host/Environment (please complete the following information):**
 - Jenkins Build Server

**Additional context**

### PR DESCRIPTION
**Pull request by mch2**
### Description
This PR is a combination of 3 changes to fix flakiness with SegmentReplicationIT.testReplicaHasDiffFilesThanPrimary.

1. Fixes a rare case a replica can be left stale if its primary does not publish a new checkpoint after the replica completes recovery.  The replica will perform a forced round of segment replication as part of the recovery process to catch up, however if it receives a new checkpoint during that time it will be dropped. If subsequently the primary stops indexing/writing new segments, the replica is left stale.  Fixed by triggering a round of segrep after the shard becomes active, if it has received a checkpoint during that time.
2. Fix case where test would query for replica's shard state while it is unavailable, resulting in "all shards failed".  Fixed this by ensuringGreen prior to performing queries.
3. Fix case where assertions trip in RefreshMetricUpdater refresh listener where AfterRefresh is invoked without a call to beforeRefresh.  This happens when an api executed refresh() is invoked on a closed NRTReplicationEngine.

### Related Issues
Resolves #6885 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java [hunk 1]
```diff
@@ -191,16 +210,18 @@ public SegmentReplicationState getSegmentReplicationState(ShardId shardId) {
      */
     public synchronized void onNewCheckpoint(final ReplicationCheckpoint receivedCheckpoint, final IndexShard replicaShard) {
         logger.trace(() -> new ParameterizedMessage("Replica received new replication checkpoint from primary [{}]", receivedCheckpoint));
+        // if the shard is in any state
+        if (replicaShard.state().equals(IndexShardState.CLOSED)) {
+            // ignore if shard is closed
+            logger.trace(() -> "Ignoring checkpoint, Shard is closed");
+            return;
+        }
+        updateLatestReceivedCheckpoint(receivedCheckpoint, replicaShard);
```
Comment (dreamer-89) [inline server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:218]: nit: Can you add an integration test verifying the fix ? One way would be to block recovery on replica while primary performs indexing.

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java [hunk 2]
```diff
@@ -152,6 +169,7 @@ public void beforeIndexShardClosed(ShardId shardId, @Nullable IndexShard indexSh
     public void shardRoutingChanged(IndexShard indexShard, @Nullable ShardRouting oldRouting, ShardRouting newRouting) {
         if (oldRouting != null && oldRouting.primary() == false && newRouting.primary()) {
             onGoingReplications.cancelForShard(indexShard.shardId(), "shard has been promoted to primary");
+            latestReceivedCheckpoint.remove(indexShard.shardId());
```
Comment (dreamer-89) [inline server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:172]: 👍

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java (hunk 1 above)
Comment (mch2) [reply server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:218]: Added test in latest commit.

**Comment by mch2** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** UNSTABLE ❕
> * **TEST FAILURES:**
> 
> ```
>       1 org.opensearch.indices.replication.SegmentReplicationIT.testScrollWithOngoingSegmentReplication
> ```
> 
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/15420/
> * **CommitID:** [510469f](https://github.com/opensearch-project/OpenSearch/commit/510469ff88ec297a4d75b61913b073a8bf987320)
>   Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

https://github.com/opensearch-project/OpenSearch/issues/7569

File: server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java [hunk 3]
```diff
@@ -337,6 +337,7 @@ public void refresh(String source) throws EngineException {
 
     @Override
     public boolean maybeRefresh(String source) throws EngineException {
+        ensureOpen();
```
Comment (dreamer-89) [inline server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java:340]: This check can be omitted as the call to `maybeRefresh()` performs this check.

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java [hunk 4]
```diff
@@ -57,43 +57,16 @@ public class SegmentReplicationTargetService implements IndexEventListener {
     private final ThreadPool threadPool;
     private final RecoverySettings recoverySettings;
 
-    private final ReplicationCollection<SegmentReplicationTarget> onGoingReplications;
+    protected final ReplicationCollection<SegmentReplicationTarget> onGoingReplications;
```
Comment (dreamer-89) [inline server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:60]: Can we defer relaxing this scope until there is a need to.

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java [hunk 5]
```diff
@@ -57,43 +57,16 @@ public class SegmentReplicationTargetService implements IndexEventListener {
...
 
     private final Map<ShardId, SegmentReplicationTarget> completedReplications = ConcurrentCollections.newConcurrentMap();
 
     private final SegmentReplicationSourceFactory sourceFactory;
 
-    private final Map<ShardId, ReplicationCheckpoint> latestReceivedCheckpoint = ConcurrentCollections.newConcurrentMap();
+    protected final Map<ShardId, ReplicationCheckpoint> latestReceivedCheckpoint = ConcurrentCollections.newConcurrentMap();
```
Comment (dreamer-89) [inline server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:66]: Same as above

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java [hunk 6]
```diff
@@ -140,8 +131,20 @@ public SegmentReplicationTargetService(
...
+        }
+    }
+
+    /**
+     * Replay any received checkpoint while replica was recovering.  This does not need to happen
+     * for primary relocations because they recover from translog.
+     */
+    @Override
+    public void afterIndexShardStarted(IndexShard indexShard) {
+        if (indexShard.indexSettings().isSegRepEnabled() && indexShard.routingEntry().primary() == false) {
```
Comment (dreamer-89) [inline server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:146]: nit: Do we need `indexShard != null` check here ?

File: server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java (hunk 3 above)
Comment (mch2) [reply server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java:340]: These are two separate methods both named `ensureOpen`, the one added here checks to ensure the engine is open, defined in Engine.java.  The one from readerManager.maybeRefresh() is on the ReferenceManager which ensures the class it is refcounting, the reader, is not closed.

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java (hunk 6 above)
Comment (mch2) [reply server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:146]: This is invoked through a composite listener [here](https://github.com/opensearch-project/OpenSearch/blob/9bf99b4a63de0e14811c0aeab0e9f854591f72ed/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L733) from within an initialized IndexShard's updateShardState, where the shard can't be null.

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java (hunk 5 above)
Comment (mch2) [reply server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:66]: This one is protected for unit test only `testUpdateLatestReceivedCheckpoint`

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java (hunk 4 above)
Comment (mch2) [reply server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java]: Done, overloaded the constructor here to pass this in for unit tests instead.

File: server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java (hunk 3 above)
Comment (dreamer-89) [reply server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java:340]: My bad, you are right.

**Review by dreamer-89 — APPROVED** [review]

**Pull request merged**